presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.type;
+
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+
+import java.util.Objects;
+
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static java.lang.String.format;
+import static java.util.Collections.singletonList;
+
+public final class CharType
+        extends AbstractVariableWidthType
+{
+    public static final int MAX_LENGTH = 65_536;
+
+    public static CharType createCharType(long length)
+    {
+        return new CharType(length);
+    }
+
+    private final int length;
+
+    private CharType(long length)
+    {
+        super(
+                new TypeSignature(
+                        StandardTypes.CHAR,
+                        singletonList(TypeSignatureParameter.of(length))),
+                Slice.class);
+
+        if (length < 1 || length > MAX_LENGTH) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, format("CHAR length scale must be in range [1, %s]", MAX_LENGTH));
+        }
+        this.length = (int) length;
+    }
+
+    public int getLength()
+    {
+        return length;
+    }
+
+    @Override
+    public boolean isComparable()
+    {
+        return true;
+    }
+
+    @Override
+    public boolean isOrderable()
+    {
+        return true;
+    }
+
+    @Override
+    public Object getObjectValue(ConnectorSession session, Block block, int position)
+    {
+        if (block.isNull(position)) {
+            return null;
+        }
+
+        return block.getSlice(position, 0, block.getLength(position)).toStringUtf8();
+    }
+
+    @Override
+    public boolean equalTo(Block leftBlock, int leftPosition, Block rightBlock, int rightPosition)
+    {
+        int leftLength = leftBlock.getLength(leftPosition);
+        int rightLength = rightBlock.getLength(rightPosition);
+        if (leftLength != rightLength) {
+            return false;
+        }
+        return leftBlock.equals(leftPosition, 0, rightBlock, rightPosition, 0, leftLength);
```

Comment:
So you are implying that if `type a` can be type only coerced to `type b` then in join operator `equalsTo` from either `type a` or `type b` would be used to compare block values (even though `type a != type b`)? I think that shouldn't be the case. If `type a` is type only coercible to `type b` that doesn't mean that `equalsTo` for those two types behaves identically (e.g coercion matters).

Consequently, this breaks VARCHAR-like comparison semantics for `CHAR` in JOINs. JOIN on `col1 CHAR(x=3)` and `col2 CHAR(y=5)` currently behaves like it first casts `col1` to `CHAR(5)`. For example:
```
presto> select * from (VALUES (CAST('a' AS CHAR(1)))) as t(a) JOIN (VALUES (CAST('a' AS CHAR(2)))) as t2(b) ON a = b;;
 a | b  
---+----
 a | a  
(1 row)
```
while:
```
presto> select * from (VALUES (CAST('a' AS CHAR(1)))) as t(a) JOIN (VALUES (CAST('a' AS CHAR(2)))) as t2(b) ON a = b WHERE a = b;
 a | b 
---+---
(0 rows)
```

I think `equalTo`, `hash` and `compareTo` that operate on `Blocks` from different types should be moved from Types in SPI and registered in `FunctionRegistry`. What do you think?

FYI: @losipiuk @fiedukow